# SV_detection_pipeline
Structural variation (SV) detection pipeline which uses DELLY, dysgu and Manta to detect SVs and report SVs which are supported by the three tools via 0.8 minimum reciprocal overlap.
## Dependencies
* Nextflow
* pyenv
* DELLY 0.7.7
* dysgu 1.3.6
* Manta 1.6.0
* BCFTools
* BEDTools
## Inputs
A reference file, indexed (using Samtools) BAM files for the samples, list of chromosomes.

List of chromosomes should be a text file:
```
Chr01
Chr02
Chr03
Chr04
Chr05
```
## Usage
Copy the pipeline script into the directory where you want to execute it. Following command is used for execution; adjust for the Python version containing dysgu installation and number of nodes:

```
nextflow sv_pipeline.nf \
--reference '/absolute/path/to/reference.fa' \
--bam_dir '/absolute/path/to/indexed-bam-files_directory/' \
--chr_list '/absolute/path/to/chr_list.txt' \
--pyenv_v 3.8.1 \
--threads 8
```
## Outputs
The pipeline will produce four VCF files. One generated by each SV-caller and the Manta-generated VCF file filtered to only contain SVs supported by all three callers via reciprocal overlap of 0.8:
* delly.vcf
* dysgu.vcf
* manta.vcf
* manta_overlap.vcf

The latter two VCF files will likely contain duplicates which can be merged:
```
bcftools norm -d all manta_overlap.vcf
```
